web/src/pages/api/public/media/upload-url.ts

```diff
@@ -0,0 +1,225 @@
+import { randomUUID } from "crypto";
+
+import { env } from "@/src/env.mjs";
+import { getFileExtensionFromContentType } from "@/src/features/media/server/getFileExtensionFromContentType";
+import { getMediaStorageServiceClient } from "@/src/features/media/server/getMediaStorageClient";
+import {
+  GetMediaUploadUrlQuerySchema,
+  GetMediaUploadUrlResponseSchema,
+  type MediaContentType,
+} from "@/src/features/media/validation";
+import { createAuthedAPIRoute } from "@/src/features/public-api/server/createAuthedAPIRoute";
+import { withMiddlewares } from "@/src/features/public-api/server/withMiddlewares";
+import {
+  ForbiddenError,
+  InternalServerError,
+  InvalidRequestError,
+} from "@langfuse/shared";
+import { prisma } from "@langfuse/shared/src/db";
+
+export default withMiddlewares({
+  POST: createAuthedAPIRoute({
+    name: "Get Media Upload URL",
+    bodySchema: GetMediaUploadUrlQuerySchema,
+    responseSchema: GetMediaUploadUrlResponseSchema,
+    successStatusCode: 201,
+    fn: async ({ body, auth }) => {
+      if (auth.scope.accessLevel !== "all") throw new ForbiddenError();
+
+      const { projectId } = auth.scope;
+      const {
+        contentType,
+        contentLength,
+        sha256Hash,
+        traceId,
+        observationId,
+        field,
+      } = body;
+
+      if (contentLength > env.LANGFUSE_S3_MEDIA_MAX_CONTENT_LENGTH)
+        throw new InvalidRequestError(
+          `File size must be less than ${env.LANGFUSE_S3_MEDIA_MAX_CONTENT_LENGTH} bytes`,
+        );
+
+      const { mediaId, uploadUrl } = await prisma.$transaction(async (tx) => {
```

Comment:
Does everything need to happen within one transaction here? Especially with the presigned link generation we may hold a long lock on the table which may lead to issues for active users.

Reply:
Great point, I think it is indeed wise to split this up 🙏🏾